Server status indicator

dit7ya opened the pull request:
## Description

This PR adds a server status (healthy | poor | critical) indicator to the user navigation bar.

**Notes to reviewers**

After halfway through this PR I realized this is a complicated one - and merits to be split into a couple of separate ones. It is NOT DONE yet but it would be very helpful if you point out if I am doing something horribly wrong.

Also, It is not clear to me yet how to best test this without manually editing stuff. But I will think about that and write here. Meanwhile, for checking out the UI component you can manually edit the initial state of the network to be `poor` and `critical` here

https://github.com/joincolony/colonyDapp/blob/db5ce6ab889116e857547b3c877c2f40cb6373f9/src/modules/pages/components/RouteLayouts/UserNavigation.tsx#L149-L151

Many thanks to Raul and Arren for helpful pointers.


NOTE: The IPFS and Reputation Oracle being behind checks will be added in a separate PR.
**New stuff** ✨

- Server status bar

Resolves #3230.

dit7ya committed:
Feat: add skeleton for server status indicator

WIP: skeleton for sever status UI

WIP: more progress on the LoaderSpinner

WIP: make UI look good

WIP: refactor a little bit

fix: make network status icons look good

dit7ya commented:
@rdig There are still some things that need to be done on this. But this is ready for review as of now so that I can have some comments/ feedback before moving on with the rest of it. I will change the commit messages with a refactor and rebase.

rdig commented:
Great, I'll review it then

rdig commented on src/modules/pages/components/RouteLayouts/UserNavigation.tsx:
```diff
@@ -115,8 +145,177 @@ const UserNavigation = () => {
+    'healthy',
+  );
+
+  const networkCheckInterval = 30 * 1000; // 15 seconds
```
Interval should be higher, at least 1 to 2 minutes, having it every 30 seconds is very expensive.

You need to strike a balance between usefulness _(of the feature)_, and load on the servers.

Also, the comment is wrong

rdig commented on src/modules/pages/components/RouteLayouts/UserNavigation.tsx:
```diff
@@ -115,8 +145,177 @@ const UserNavigation = () => {
+          reputationOracleResponsePromise,
+        ]);
+
+        if (rpcBlockResponse.status !== 200) {
+          setNetworkHealth('critical');
+          return;
+        }
```
A lot of services might take more than 10 seconds to respond _(due to other network factors)_, so I think this will trigger even for services that are running.

rdig commented on src/modules/pages/components/RouteLayouts/UserNavigation.tsx:
```diff
@@ -115,8 +145,177 @@ const UserNavigation = () => {
     <div className={styles.main}>
+      {networkHealth !== 'healthy' && (
+        <div>
+          <MiniSpinnerLoaderWrapper milliseconds={networkHealthLoadingTime}>
```
Why is this decoupled in terms of logic?

Why not hide it when everything is set up properly here ?

rdig reviewed (commented):
These are just some notes I observed at a first pass.

There's stuff you should change, but the general idea you are going for is pretty good.

danbr commented on src/modules/core/components/MiniSpinnerLoaderWrapper/MiniSpinnerLoaderWrapper.tsx:
```diff
@@ -0,0 +1,27 @@
+  milliseconds: number;
+}
+
+const MiniSpinnerLoaderWrapper = ({ children, milliseconds }: Props) => {
```
I think you could choose a better name for the component. Being specific about what it has/does more than `MiniSpinnerLoader` (not more general).

Also `milliseconds` is a wasted opportunity of a variable name. It doesnt tell me whats it being used for, or give a context. 
A unit type as a name could be confusing.

dit7ya replied:
Good points. I will change `milliseconds` to something like `loadingDuration`. Any suggestion for a better component name?

danbr replied:
> Good points. I will change milliseconds to something like loadingDuration. Any suggestion for a better component name?

I'm trying to improve my own naming convention thought process.  I need to daydream for a while...... My default is usually to be very literal. 

Probably `MiniSpinnerLoader` isnt a great name (Mini) to start with, as the appearance can easily be changed with a prop. But now you have to build upon it!!! (or start changing lots of things)

Do we even need another `Mini-SpinnerLoader-Wrapper` component? We have lots and they are all very similar. 

I think that having a timeout is a useful feature in a spinner. 
So you should not make another component but instead, add the timeout to the 'base' component SpinnerLoader. Add the `loadingDuration` prop (default=0) then everyone can make use of it, or not.

..... and for the extra work, you get to avoid the problem of deciding on an expressive name ;)

danbr commented on src/modules/pages/components/RouteLayouts/UserNavigation.css:
```diff
@@ -190,3 +190,48 @@
+  margin-right: 8px;
+}
+
+.networkHealthIcon-critical svg {
```
I'm interested why you opted for kebab-case? I thought we should use only camelcase? (even though the *.css.d.ts generates camelcase)

dit7ya replied:
I saw other instances of kebab-case in the codebase. For example - https://github.com/joincolony/colonyDapp/blob/41f9ed47667ca73822740cf4e04b88ed0f4b00f5/src/modules/core/components/OmniPicker/OmniPicker.tsx#L227 .

danbr replied:
👍 I'll try to use it if an appropriate situation arises.

arrenv reviewed (commented):
My thoughts are that it would be good to separate all this logic out of `UserNavigation` into it's own component. What do you think?

In terms of design, while the loader colour is different, I think the grey is fine to be consistent with the rest of the loaders. However, it needs some right margin.

![13a3e0a6-6800-417c-887a-9be61822fa52](https://user-images.githubusercontent.com/33682027/162430406-39a5466f-000b-4bf0-a378-d71a849db1f0.png)

Also, the icons, look a little bold, you can make the stroke transparent to improve `stroke: transparent;`

![579ffc91-00b4-4f48-a465-254f48db97f1](https://user-images.githubusercontent.com/33682027/162430684-1cc76877-47aa-4937-8dda-2860580b4045.png)

dit7ya committed:
Fix: refactor to use canonical state management

dit7ya committed:
Fix: refactor out NetworkStatus component

dit7ya committed:
Fix: remove the SpinnerLoader altogether

This was decided in a meeting.

Fix: remove MiniSpinnerLoaderWrapper

Fix: position the Spinner correctly

Feat: add condition for critical state

Fix: refactor and clean up

Fix: remove the SpinnerLoader altogether

rdig commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,176 @@
+                    </span>
+                  </div>
+                  <span className={styles.networkHealthDescription}>
+                    {formatMessage(MSG.networkHealthDescription, {
+                      networkHealth,
+                    })}
```
Same as above

rdig commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,176 @@
+              }}
+            >
+              <div
+                className={`${styles.elementWrapper}
+                            ${styles.networkInfo}
+                            ${styles.networkHealthHeading}`}
```
What triggered this code styling here? Why not have them inlined ?

dit7ya replied:
Do you mean why not put these three lines in a single line? It exceeds the 80 character line length limit enforced by the linter.

rdig commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,176 @@
+                >
+                  <Icon name="triangle-warning" />
+                </span>
+                Network
```
This needs to be a message descriptor

rdig reviewed (changes requested):
You're heading in the right direct, however there's still a lot of changes needed before this is ready.

I've left notes on what you should change

dit7ya commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,176 @@
+  const loadingDuration = 1 * 1000;
+
+  useEffect(() => {
+    const loadingTimer = setTimeout(() => {
+      setShowLoading(false);
+    }, loadingDuration);
+    return () => clearTimeout(loadingTimer);
+  }, [loadingDuration]);
```
This is the intended behavior per the spec. The network status indicator is only visible when it is the status is not healthy, and the 1 second loading spinner is just cosmetic.

https://www.figma.com/file/zGguqh4iZSIdzBZZMylKYn/Improve-Network-Health?node-id=258%3A8700

rdig replied:
That spec doesn't make sense. 

What's "start to boil?" Are we to detect if the server is "about to go down", that's really not possible with the current infrastructure.

So either the spec needs to change, or it needs to be clarified.

But at any rate, that 1 second spinner, with no actual logic tied to it, is not something we want to add.

rdig commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,185 @@
+    pollInterval: networkCheckInterval,
+  });
+
+  useEffect(() => {
+    const networkCheck = setInterval(async () => {
+      if (latestRpcBlockError) {
+        setNetworkHealth(NetworkHealthType.Critical);
+      } else if (
+        latestSubgraphBlockError ||
+        !isReputationOracleAlive?.isReputationOracleAlive ||
+        !isColonyServerAlive?.isServerAlive ||
+        (latestRpcBlock &&
+          latestSubgraphBlock &&
+          latestRpcBlock.latestRpcBlock >
+            latestSubgraphBlock.latestSubgraphBlock)
+      ) {
+        setNetworkHealth(NetworkHealthType.Poor);
+      } else {
+        // If everything is okay, set health to healthy (to correct for the previous state)
+        setNetworkHealth(NetworkHealthType.Healthy);
+      }
+    }, networkCheckInterval);
+    return () => clearInterval(networkCheck);
+  });
```
Running these every second incurs a very serious performance penalty. Both in app memory, but especially in infrastrcture resources.

See demo: 
![demo-1m-requests-per-second](https://user-images.githubusercontent.com/1193222/163832261-6232eb6e-54f7-40dd-8a61-93ab68dd35d3.gif)

This is not necessarily a problem with your implementation as you just followed the specs, and more of a problem of specs not looking at the big picture.

For this we need to take a step back and come up with some sort of deffering or debouncing mechanism _(although debouncing is not really suitable here)_

We need to have a call with the Product team and come up with alternatives as I'm not comfortable pushing this into production.

Make sure to schedule a call w/ you, @arrenv and me to hash this out.

rdig reviewed (changes requested):
This is mostly fine, however, I think the specs need to be amended as they aren't production worthy the way they are right now.

We need to get to the bottom of two things:
- that spinner logic
- the excessive strain on our infrastructure

Please address the minor notes I've left, but you can consider this blocked until we find a better alternative. 

_(make sure to schedule that call I've mentioned in which we'll discuss these things)_

rdig commented:
@dit7ya Please don't forget to set up that call w/ Product

dit7ya committed:
Fix: change networkCheckInterval to 5 minutes

dit7ya committed:
Fix: move NetworkStatus queries to its own file

arrenv commented on src/data/graphql/queries/status.graphql:
```diff
@@ -0,0 +1,16 @@
+query LatestRpcBlock {
+  latestRpcBlock @client
+}
+
+
```
Same comment as Raul, an extra space here.

arrenv reviewed (approved):
From what I can tell, it looks good! Nice work on this.
I definitely think Raul should do a review of this one again, based off of our discussions. You've pinged him though, so all good.

dit7ya committed:
Fix: move out type declaration and small lint

Fix: remove empty line

Fix: move enum out and small lint

Fix: move out type declaration to separate file

chinins reviewed (approved):
Looks good. I am just curious why you decided to go for 4 queries instead of 1 that combines 4 values (I saw Raul's comment that there is a reason for this).

dit7ya commented:
@chinins The rationale for splitting the queries is that we might be interested in the individual loading or error states of the requests in this case.

rdig commented on src/modules/core/components/NetworkStatus/NetworkStatus.tsx:
```diff
@@ -0,0 +1,169 @@
+
+const NetworkStatus = () => {
+  enum NetworkHealthType {
+    Healthy = 'healthy',
```
I think what Olga meant here is to move them outside of the component **file** as well, and just import them, usually inside `src/types`

That way they can be share by the whole dapp, in case there's a need in the future to touch, or integrate, parts of the server health system somewhere else

rdig reviewed (approved):
Tested one last time and I can confirm that the specific requests the status component makes are only fired once every 5 minutes.

We do make a shit ton of other requests though, which we'll need to cut down in the future version of CDapp

ArmandoGraterol committed:
Added isIPFSAlive query

ArmandoGraterol committed:
Updated isReputationOracleAlive and isIPFSAlive

ArmandoGraterol committed:
Added isIPFSAlive to NetworkStatus

ArmandoGraterol committed:
Added extra check for isReputationOracleAlive for dev use case

ArmandoGraterol committed:
Removed unused setInterval from useEffect in NetworkStatus

dit7ya committed:
Fix: rename to popperOptions in NetworkStatus

dit7ya committed:
Fix: use specific hash in isIpfsAlive

dit7ya committed:
Feat: add console info for NetworkStatus changes

dit7ya committed:
Feat: move subgraph behind condition by 42 blocks

dit7ya committed:
Fix: remove console.info statements

dit7ya committed:
Merge pull request #3415 from JoinColony/fix/3408-isIpfsAlive-use-hash

Use specific hash in isIpfsAlive and move subgraph check behind

arrenv commented:
Does anyone remember the status of this one?

ArmandoGraterol commented:
@arrenv 
> Does anyone remember the status of this one?

I remember that this feature went out to QA and it was difficult to test because we basically needed to manually sabotage our services in order to see changes in the server statuses.

That's the only I can recall, perhaps someone else remembers more

rdig commented:
There was still something to fix regarding IPFS if I remember correctly, but I haven't had time to look into it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/JoinColony/colonyDapp?pullRequest=3259) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/JoinColony/colonyDapp?pullRequest=3259) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ArmandoGraterol<br/>:x: dit7ya<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/JoinColony/colonyDapp?pullRequest=3259) it.</sub>
